### PR TITLE
added support for regex path and rewrite in proxy options

### PR DESCRIPTION
**Pull request by nthtran**
Fix of #148.

### CODE REVIEW

**Comment by rafaelchiti** [comment]
@sokra is there intentions to merge this in? just to know since I'll be needing it. 
Thanks!

**Comment by wmertens** [comment]
This is great, especially the host option, exactly what I needed!

@sokra please merge :grin:

File: bin/webpack-dev-server.js [hunk 1]
```diff
@@ -140,17 +140,14 @@ new Server(webpack(wpOpt), options).listen(options.port, options.host, function(
...
-			} else {
-				target = options.proxy[path].target;
-			}
-			console.log('proxying '+path+' to '+target);
+	if (options.proxy && !Array.isArray(options.proxy)) {
+		options.proxy = Object.keys(options.proxy).map(function (path) {
+			return {
+				path: path,
+				target: typeof options.proxy[path] === "string" ? options.proxy[path] : options.proxy[path].target
+			};
```
Comment (sokra) [inline bin/webpack-dev-server.js]: This could be removed, because lib/Server.js does the same.

File: lib/Server.js [hunk 2]
```diff
@@ -110,15 +110,21 @@ function Server(compiler, options) {
...
-				proxyOptions = options.proxy[path];
-			}
-			app.all(path, function (req, res) {
+		if (!Array.isArray(options.proxy)) {
+			options.proxy = Object.keys(options.proxy).map(function (path) {
+				var proxyOptions;
+				if (typeof options.proxy[path] === 'string') {
+					proxyOptions = {path: path, target: options.proxy[path]};
+				} else {
+					proxyOptions = options.proxy[path];
```
Comment (sokra) [inline lib/Server.js]: `proxyOptions.path = path;` missing?

File: bin/webpack-dev-server.js (hunk 1 above)
Comment (nthtran) [inline bin/webpack-dev-server.js]: Remove the whole `if(options.proxy...)`?

File: lib/Server.js (hunk 2 above)
Comment (nthtran) [inline lib/Server.js]: Added.

File: bin/webpack-dev-server.js (hunk 1 above)
Comment (sokra) [inline bin/webpack-dev-server.js]: It's not necessary, is it?

Comment (nthtran) [inline bin/webpack-dev-server.js]: Yeah it's just for `console.log`. It's straight from #148, I'll just remove it then.

Comment (nthtran) [inline bin/webpack-dev-server.js]: Removed.

**Comment by sokra** [comment]
Thanks

**Comment by jokull** [comment]
Should be noted this isn’t real regex support but [path-to-regexp](https://github.com/pillarjs/path-to-regexp) support. Took me a while to figure out :ghost:

**Comment by sokra** [comment]
@jokull You can send a PR to add these hint...

**Comment by kisenka** [comment]
Sorry, can you explain how to use this feature?

**Comment by nthtran** [comment]
@kisenka Please see #148 for example usage. There are more examples in #127 too.

**Comment by stevehu** [comment]
I want to proxy to my API server if uri is /api/rs to http://anotherhost:8080/api/rs. What would be my proxy configuration? I tried several examples and they are not working. Any help would be greatly appreciated. Some people post solutions that use express server to proxy to webpack-dev-server and I don't want to go that route. Thanks.

**Comment by stevehu** [comment]
Sorry. Please ignore my previous post as it is working now after upgrade to the latest webpack and webpack-dev-server. Thanks a lot for fixing it.

**Pull request merged**